### PR TITLE
Implement QLinear GlobalAveragePool with sse2/neon.

### PR DESCRIPTION
**Pull request by zhanghuanrong**
Also start adding google test into mlas test.

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include "core/util/math.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+class QLinearGlobalAveragePool final : public OpKernel {
+ public:
+  QLinearGlobalAveragePool(const OpKernelInfo& info) : OpKernel(info) {
+    int64_t nchw_layout;
+    ORT_ENFORCE(info.GetAttr("nchw", &nchw_layout).IsOK());
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.h:17]: For consistency, make this be a channels_last<int64_t> attribute like com.ms.QLinearConv and ReorderOutput.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -1992,6 +1992,95 @@ with the exception that numpy default keepdims to False instead of True.)DOC")
       .FillUsing(QLinearMathDocGenerator("multiplication",
                                          "C = ((A - A_zero_point) * (B - B_zero_point)) * (A_scale * B_scale)/C_scale + C_zero_point"));
 
+  ONNX_CONTRIB_OPERATOR_SCHEMA(QLinearGlobalAveragePool)
```
Comment (tracysh) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:1995]: This schema can go into nhwc_defs.cc now.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -1992,6 +1992,95 @@ with the exception that numpy default keepdims to False instead of True.)DOC")
...
+      .SetDomain(kMSDomain)
+      .SinceVersion(1)
+      .SetDoc(R"DOC(
+QLinearGlobalAveragePool consumes an input tensor X and applies Average pooling across
+the values in the same channel. This is equivalent to AveragePool with kernel size
+equal to the spatial dimension of input tensor. Input is of type uint8_t or int8_t.
+)DOC")
+      .Attr(
+          "nchw",
```
Comment (tracysh) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2004]: change to channels_last per other comment

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 4]
```diff
@@ -1992,6 +1992,95 @@ with the exception that numpy default keepdims to False instead of True.)DOC")
...
+          "Input layout, nchw or nhwc, 1 for nchw, 0 for nhwc.",
+          AttributeProto::INT)
+      .Input(
+          0,
+          "X",
+          "Input data tensor from the previous operator; "
```
Comment (tracysh) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2010]: I was lazy and didn't put these comments into com.ms.QLinearConv, but if you need to keep them, the part about the image size is not accurate if you have channels last.

File: onnxruntime/core/mlas/inc/mlas.h [hunk 5]
```diff
@@ -709,6 +709,43 @@ MlasFindMinMaxElement(
...
+size_t
+MLASCALL
+MlasQLinearSafePaddingElementCount(
+    size_t ElementSize,
+    size_t ElementCount
+    );
+
+void
+MLASCALL
+MlasQLinearGlobalAveragePool(
```
Comment (tracysh) [inline onnxruntime/core/mlas/inc/mlas.h:721]: Rename to MlasQLinearGlobalAveragePoolNchw

File: onnxruntime/core/mlas/inc/mlas.h [hunk 6]
```diff
@@ -709,6 +709,43 @@ MlasFindMinMaxElement(
...
+    uint8_t* Output,
+    float ScaleOutput,
+    int32_t ZeroPointOutput,
+    size_t Channels,
+    size_t ImageSize
+    );
+
+void
+MLASCALL
+MlasNhwcQLinearGlobalAveragePool(
```
Comment (tracysh) [inline onnxruntime/core/mlas/inc/mlas.h:734]: Rename to MlasQLinearGlobalAveragePoolNhwc (same pattern as MlasReorderOutputXxx)

File: onnxruntime/core/mlas/lib/quantize.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        *dst++ = *src++;
+        --N;
+    }
+    return target;
+}
+
+#if defined(MLAS_SSE2_INTRINSICS)
+
+MLAS_FORCEINLINE
+static MLAS_INT32X4
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.h:73]: remove static

File: onnxruntime/test/mlas/unittest.cpp [hunk 8]
```diff
@@ -22,6 +22,8 @@ Module Name:
 #include <random>
 #include <mlas.h>
 
+#include "gtest/gtest.h"
```
Comment (tracysh) [inline onnxruntime/test/mlas/unittest.cpp:25]: Are the gtest changes needed as part of this PR? Would rather see that split out as you aren't really using it much here.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+          tp, static_cast<std::ptrdiff_t>(N),
+          {1.0 * kernel_size * C, 1.0 * C, 8.0 *kernel_size * C},
+          worker);
+    }
+  }
+  return Status::OK();
+}
+
+Status QLinearGlobalAveragePool::Compute(OpKernelContext* context) const {
+  auto tensor_x_scale = context->Input<Tensor>(1);
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:77]: use "const auto*" here to be clearer?

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+  } else {
+    output_dims.insert(output_dims.end(), one_dims.begin(), one_dims.end());
+    output_dims.push_back(C);
+  }
+  Tensor& Y = *context->Output(0, output_dims);
+
+  const float x_scale = *(tensor_x_scale->Data<float>());
+  const float y_scale = *(tensor_y_scale->Data<float>());
+  auto dtype = X.GetElementType();
+  switch (dtype) {
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:115]: Is this dispatch smaller than registering the kernel explicitly for uint8_t type instead?

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  const auto& X = *context->Input<Tensor>(0);
+  const auto& x_shape = X.Shape().GetDims();
+
+  ORT_RETURN_IF_NOT(x_shape.size() >= 3, "Input dimension cannot be less than 3.");
+  auto first_dim = x_shape.begin() + (storage_order_ == StorageOrder::NCHW ? 2 : 1);
+  std::vector<int64_t> kernel_dims{first_dim, first_dim + (x_shape.size() - 2)};
+  int64_t N = x_shape[0];
+  int64_t C = (storage_order_ == StorageOrder::NCHW ? x_shape[1] : x_shape.back());
+
+  std::vector<int64_t> output_dims{N};
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:101]: Would the code be smaller to copy x_shape to output_dims and then just set first_dim to first_dim + kernel_rank to 1? One less std::vector to construct plus not all the push_back calls.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+              "input y_scale must be a scalar or 1D tensor of size 1");
+  ORT_ENFORCE(IsScalarOr1ElementVector(tensor_y_zero_point),
+              "input y_zero_point must be a scalar or 1D tensor of size 1 if given");
+
+  concurrency::ThreadPool* tp = context->GetOperatorThreadPool();
+  const auto& X = *context->Input<Tensor>(0);
+  const auto& x_shape = X.Shape().GetDims();
+
+  ORT_RETURN_IF_NOT(x_shape.size() >= 3, "Input dimension cannot be less than 3.");
+  auto first_dim = x_shape.begin() + (storage_order_ == StorageOrder::NCHW ? 2 : 1);
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:96]: woluld be nice to call this spatial_dim_start to match QLinearConv implementation style

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            Input, x_scale, x_zero_point, Output, y_scale, y_zero_point,
+            N, kernel_size, C, channel_count, acc_buffer.data(), zero_buffer.data());
+      };
+      concurrency::ThreadPool::TryParallelFor(
+          tp, static_cast<std::ptrdiff_t>(channel_groups),
+          {1.0 * N * kernel_size * kMiniChannelGroup, 1.0 * N * kMiniChannelGroup, 8.0 * N * kernel_size * kMiniChannelGroup},
+          worker);
+    } else {
+      auto worker = [=](std::ptrdiff_t first, std::ptrdiff_t last) {
+        const uint8_t* Input = x + first * C * kernel_size;
```
Comment (tracysh) [inline onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:59]: Input->input, Output->output?

File: onnxruntime/core/mlas/lib/quantize.h [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+#include "mlasi.h"
+
+union MLAS_FLOAT32BITS {
+    uint32_t u32;
+    float fp32;
+};
+
+MLAS_FORCEINLINE
+static uint32_t
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.h:28]: static shouldn't be necessary (throughout file)?

File: onnxruntime/core/mlas/lib/qlgavgpool.cpp [hunk 15]
```diff
@@ -0,0 +1,815 @@
+/*++
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Licensed under the MIT License.
+
+Module Name:
+
+    quantize.cpp
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/qlgavgpool.cpp:9]: fix module name/description

**Review by tracysh — CHANGES_REQUESTED** [review]

File: onnxruntime/core/mlas/lib/qlgavgpool.cpp [hunk 16]
```diff
@@ -0,0 +1,815 @@
...
+    const uint8_t* i3 = i2 + Stride;
+    const uint8_t* i6 = i5 + Stride;
+
+    for (; ImageSize > 7; ImageSize -= 7) {
+        int32_t* acc = AccumulateBuffer;
+        size_t c = Channels;
+        for (; c >= 8; c -= 8) {
+            LOAD_FULL_CHANNELS();
+
+            CACULATE_ACCUMULATE_VECTORS();
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/qlgavgpool.cpp:330]: CACULATE->CALCULATE

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 4 above)
Comment (zhanghuanrong) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:2010]: Right. Add C last case.

File: onnxruntime/core/mlas/inc/mlas.h (hunk 6 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/inc/mlas.h:734]: Done

File: onnxruntime/core/mlas/lib/quantize.h (hunk 7 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/quantize.h:73]: Yeah, it is inside SSE2.

File: onnxruntime/test/mlas/unittest.cpp (hunk 8 above)
Comment (zhanghuanrong) [reply onnxruntime/test/mlas/unittest.cpp:25]: No problem. Let me sperate it in next PR.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc (hunk 10 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:115]: If the spec keep u8/s8 for future expansion, it is may be better keep it like here.
How about let keep it this way.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc (hunk 9 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:77]: yes. fixed.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc (hunk 11 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:101]: Fixed.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc (hunk 13 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc:59]: Fixed.

File: onnxruntime/core/mlas/lib/qlgavgpool.cpp (hunk 16 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/qlgavgpool.cpp:330]: Fixed.

File: onnxruntime/core/mlas/lib/qlgavgpool.cpp (hunk 15 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/qlgavgpool.cpp]: Fixed.

File: onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc (hunk 12 above)
Comment (zhanghuanrong) [reply onnxruntime/contrib_ops/cpu/qlinear_global_average_pool.cc]: fixed.

File: cmake/onnxruntime_mlas.cmake [hunk 17]
```diff
@@ -4,7 +4,6 @@
 set(mlas_common_srcs
   ${ONNXRUNTIME_ROOT}/core/mlas/lib/platform.cpp
   ${ONNXRUNTIME_ROOT}/core/mlas/lib/threading.cpp
-  ${ONNXRUNTIME_ROOT}/core/mlas/lib/dgemm.cpp
```
Comment (tracysh) [inline cmake/onnxruntime_mlas.cmake:7]: why did dgemm.cpp need to move?

File: onnxruntime/core/optimizer/nhwc_transformer.cc [hunk 18]
```diff
@@ -213,6 +214,23 @@ void NhwcTransformerImpl::TransformQLinearActivation(Node& node) {
   CreateNhwcArgument(node, node, nhwc_input->rank_);
 }
 
+void NhwcTransformerImpl::TransformQLinearGlobalAveragePool(Node& node) {
+  auto& input_defs = node.MutableInputDefs();
+
```
Comment (tracysh) [inline onnxruntime/core/optimizer/nhwc_transformer.cc:219]: You should add a check that channels_last is not already 1.

File: onnxruntime/core/mlas/lib/quantize.h [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+    while (N > 0) {
+        *dst++ = *src++;
+        --N;
+    }
+    return target;
+}
+
+#if defined(MLAS_SSE2_INTRINSICS)
+
+MLAS_FORCEINLINE
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.h:72]: This no longer needs to move, right?

File: onnxruntime/core/mlas/lib/quantize.cpp [hunk 20]
```diff
@@ -18,7 +18,8 @@ Module Name:
 
 --*/
 
-#include "mlasi.h"
+#include "quantize.h"
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.cpp:21]: This file shouldn't need to change now, right?

File: onnxruntime/core/mlas/lib/quantize.h [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+    uint32_t u
+    )
+{
+    MLAS_FLOAT32BITS uf = {u};
+    return uf.fp32;
+}
+
+MLAS_FORCEINLINE
+static uint8_t*
+MlasCopyTailBytes(
```
Comment (tracysh) [inline onnxruntime/core/mlas/lib/quantize.h:50]: Does this function need to move now?

File: cmake/onnxruntime_mlas.cmake (hunk 17 above)
Comment (zhanghuanrong) [reply cmake/onnxruntime_mlas.cmake:7]: currently windows build with vs 2017 for x86 treat it as error as empty object file generated.

File: onnxruntime/core/mlas/lib/quantize.h (hunk 19 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/quantize.h:72]: Yes. I will move it back.

File: onnxruntime/core/mlas/lib/quantize.cpp (hunk 20 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/quantize.cpp:21]: Yes.

File: onnxruntime/core/mlas/lib/quantize.h (hunk 21 above)
Comment (zhanghuanrong) [reply onnxruntime/core/mlas/lib/quantize.h:50]: No need too. Will move back this time. And remove them to memcpy in seperate PR.

File: onnxruntime/core/optimizer/nhwc_transformer.cc (hunk 18 above)
Comment (zhanghuanrong) [reply onnxruntime/core/optimizer/nhwc_transformer.cc:219]: Add check with ORT_ENFORCE.

File: onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc [hunk 22]
```diff
@@ -168,6 +168,82 @@ void RegisterNhwcSchemas() {
...
+        auto input_shape = ctx.getInputType(0)->tensor_type().shape();
+        if (input_shape.dim_size() < 2) {
+          return;
+        }
+
+        // (N, C, 1, 1, ..., 1) or (N, 1, 1, ..., 1, C)
+        auto output_shape = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+        output_shape->CopyFrom(input_shape);
+        int image_dim_index = (channel_last ? 1 : 2);
+        for (auto n_hw_dims = input_shape.dim_size() - 2; n_hw_dims > 0; --n_hw_dims) {
```
Comment (tracysh) [inline onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc:243]: If any of the spatial dimensions have symbolic values, then with this code, the dimension will have both an integer and symbolic value. To be safest with existing usage, maybe just do the individual add_dim calls to the output shape instead?

Comment (zhanghuanrong) [reply onnxruntime/core/graph/contrib_ops/nhwc_schema_defs.cc:243]: I c. Will change it clear_dim_value() first. Then set_dim_value().

**Review by tracysh — APPROVED** [review]

**Pull request merged**